pkg/eestream/decode.go

```diff
@@ -14,28 +16,59 @@ import (
 type decodedReader struct {
 	rs     map[int]io.ReadCloser
 	es     ErasureScheme
-	inbufs map[int][]byte
 	outbuf []byte
 	err    error
+	chans  map[int]chan block
+	cb     int // current block number
 }
 
 type readerError struct {
 	i   int // reader index in the map
 	err error
 }
 
+type block struct {
+	i    int    // reader index in the map
+	num  int    // block number
+	data []byte // block data
+	err  error  // error reading the block
+}
+
 // DecodeReaders takes a map of readers and an ErasureScheme returning a
 // combined Reader. The map, 'rs', must be a mapping of erasure piece numbers
 // to erasure piece streams.
 func DecodeReaders(rs map[int]io.ReadCloser, es ErasureScheme) io.ReadCloser {
```

Comment:
eventually should take a context.Context as the first argument. you can then use context.WithCancel to give yourself a way to cancel the context, then have all of the parallel readers react to context cancelation